plugins/GSdx/GSDump.cpp

```diff
@@ -22,74 +22,165 @@
+}
 
-	m_frames = 0;
-	m_extra_frames = 2;
+void GSDumpBase::Transfer(int index, const uint8* mem, size_t size)
+{
+	if (size == 0)
+		return;
 
-	if(m_gs)
-	{
-		fwrite(&crc, 4, 1, m_gs);
-		fwrite(&fd.size, 4, 1, m_gs);
-		fwrite(fd.data, fd.size, 1, m_gs);
-		fwrite(regs, sizeof(*regs), 1, m_gs);
-	}
+	AppendRawData(0);
+	AppendRawData(index);
+	AppendRawData(&size, 4);
+	AppendRawData(mem, size);
 }
 
-void GSDump::Close()
+void GSDumpBase::ReadFIFO(uint32 size)
 {
-	if(m_gs) {fclose(m_gs); m_gs = NULL;}
+	if (size == 0)
+		return;
+
+	AppendRawData(2);
+	AppendRawData(&size, 4);
 }
 
-void GSDump::Transfer(int index, const uint8* mem, size_t size)
+bool GSDumpBase::VSync(int field, bool last, const GSPrivRegSet* regs)
 {
-	if(m_gs && size > 0)
-	{
-		fputc(0, m_gs);
-		fputc(index, m_gs);
-		fwrite(&size, 4, 1, m_gs);
-		fwrite(mem, size, 1, m_gs);
-	}
+	AppendRawData(3);
+	AppendRawData(regs, sizeof(*regs));
+
+	AppendRawData(1);
+	AppendRawData(field);
+
+	if (last)
+		m_extra_frames--;
+
+	return ((++m_frames & 1) == 0 && last && (m_extra_frames < 0));
 }
 
-void GSDump::ReadFIFO(uint32 size)
+//////////////////////////////////////////////////////////////////////
+// GSDump implementation
+//////////////////////////////////////////////////////////////////////
+
+GSDump::GSDump(const string& fn, uint32 crc, const GSFreezeData& fd, const GSPrivRegSet* regs) : GSDumpBase(fn + ".gs")
 {
-	if(m_gs && size > 0)
-	{
-		fputc(2, m_gs);
-		fwrite(&size, 4, 1, m_gs);
+	AddHeader(crc, fd, regs);
+}
+
+void GSDump::AppendRawData(const void *data, size_t size)
+{
+	size_t written = fwrite(data, size, 1, m_gs);
+	if (written != size)
+		fprintf(stderr, "GSDump: Error failed to write data\n");
+}
+
+void GSDump::AppendRawData(uint8 c)
+{
+	if (fputc(c, m_gs) == EOF)
+		fprintf(stderr, "GSDump: Error failed to write data\n");
+}
+
+//////////////////////////////////////////////////////////////////////
+// GSDumpXz implementation
+//////////////////////////////////////////////////////////////////////
+
+#ifdef LZMA_SUPPORTED
+
+GSDumpXz::GSDumpXz(const string& fn, uint32 crc, const GSFreezeData& fd, const GSPrivRegSet* regs) : GSDumpBase(fn + ".gs.xz")
+{
+	m_in_buff.clear();
+
+	m_strm = LZMA_STREAM_INIT;
+	lzma_ret ret = lzma_easy_encoder(&m_strm, 6 /*level*/, LZMA_CHECK_CRC64);
+	if (ret != LZMA_OK) {
+		fprintf(stderr, "GSDumpXz: Error initializing LZMA encoder ! (error code %u)\n", ret);
+		return;
 	}
+
+	AddHeader(crc, fd, regs);
 }
 
-void GSDump::VSync(int field, bool last, const GSPrivRegSet* regs)
+GSDumpXz::~GSDumpXz()
 {
-	if(m_gs)
+	Flush(true);
+}
+
+void GSDumpXz::AppendRawData(const void *data, size_t size)
+{
+	size_t old_size = m_in_buff.size();
+	m_in_buff.resize(old_size + size);
+	memcpy(&m_in_buff[old_size], data, size);
+
+	// Enough data was accumulated, time to write/compress it.  If compression
+	// is enabled, it will freeze PCSX2. 1GB should be enough for long dump.
+	if (m_in_buff.size() > 1024*1024*1024)
+		Flush(false);
```

Comment:
I guess the if won't ever be true anymore.

Actually, unless I'm misreading something, I think there's a small risk that `lzma_code(&m_strm, LZMA_FINISH)` won't be called if this path is ever taken (VSync -> flush(false) -> empty buffer -> destructor -> flush(true) -> early return -> `lzma_code(&m_strm, LZMA_FINISH)` not called + memleak because `lzma_end()` isn't called).